### ISSUE TITLE
Atomic state invariants

### ISSUE THREAD

**Issue by barrucadu** [barrucadu/dejafu#177]
**Prerequisite: #202**

As I think setting up invariants should only happen in the "arrange" phase.

---

I am imagining something like:

```haskell
-- | Run a concurrent computation and return its result, also enforcing
-- some invariants which must hold at all scheduling points.
--
-- Invariants are checked before all scheduling points.  If an invariant
-- returns @Nothing@, it has passed.  If it returns @Just err@, the inner
-- computation is aborted with @InvariantFailure err@.  Invariants may not
-- have visible side-effects, such as emptying an @MVar@ (excluding
-- relaxed memory effects, which are ignored).  If an invariant has a visible
-- side-effect, the inner computation aborts with @IllegalInvariant@.
--
-- Like `subconcurrency`, this can only be called in the main thread,
-- when no other threads exist, and calls cannot be nested (nor can
-- `subconcurrencyWithInvariants` be called inside `subconcurrency`,
-- or vice versa).
subconcurrencyWithInvariants
  :: [ConcT r n (Maybe String)]
  -- ^ The list of invariants to check.
  -> ConcT r n a
  -- ^ The inner computation.
  -> ConcT r n (Either Failure a)

-- also:
subconcurrency = subconcurrencyWithInvariants []
```

The Haddock should be fairly self-explanatory.

---

This can be seen as a strictly weaker version of the temporal logic property-checking I implemented in the monad-monitor work.  That work was really inefficient as it used global state to keep track of which "events" were currently active (properties were about these "events").  So dejafu checked different orderings of updating this global state, which caused an explosion of schedules.

The paper was rejected, and quite rightly so.  But by adding these weaker invariants directly in dejafu we get a couple of nice properties:

- Execution exits early if an invariant is violated.
- They are atomic. You can say that the values of two `MVars` must always satisfy some property if you want.

I suspect there are a lot of properties which people care about which are of the form "this invariant should always hold".

---

Invariants must be effect-free, to avoid messing with the main computation, and to keep the order in which invariants are checked irrelevant.  But there are some remaining questions to consider:

- What happens if multiple invariants fail at once?  Should all the errors be returned?  Or should the first failing one be returned?
- What happens if the invariant blocks or throws an exception?  Is that an "invariant failure"?
- Should invariants be checked before or after each scheduling point?
- What information should be passed to an invariant?  Can they only monitor global state set up before the `subconcurrencyWithInvariants` call, or do they also get the `ThreadId` and `ThreadAction` which just happened?
- The term "failure" already means something in dejafu, so perhaps a different word should be used in this context.

**Comment by taktoa** [comment]
> This can be seen as a strictly weaker version of the temporal logic property-checking I implemented in the monad-monitor work.

This sounds like runtime verification (membership : inclusion :: runtime verification : model checking); can you link me to the "monad-monitor work"? I'm somewhat interested in writing a runtime verification library for Haskell.

**Comment by barrucadu** [comment]
Here you go: https://github.com/barrucadu/monad-monitor

It is a form of runtime verification, and I did it after being influenced by work in that area; but this particular manifestation of it doesn't work out so well.  The way I did it is to add a `TVar` containing an event log, and a bunch of functions to update it.  The problem is that dejafu then tries to find different interleavings of all these functions, and the search space explodes.

There's quite possibly a better way of doing it without adding a new primitive to dejafu.

---

Out of interest, has dejafu been mentioned anywhere recently?  I've seen four new people pop up on the github in the past couple of days, after months of me not doing anything.  I'm a little curious.
